File: app/src/main/java/dev/ycosorio/agenda/MainActivity.java
package dev.ycosorio.agenda;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.Toast;

import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Locale;

import dev.ycosorio.agenda.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;

    private EventoAdapter adapter;
    private ArrayList<Evento> listaEventos;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        ViewCompat.setOnApplyWindowInsetsListener(binding.getRoot(), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        
        binding.btnGuardar.setOnClickListener(view -> {

            String titulo = binding.etTitulo.getText().toString();
            String sFecha = binding.etFecha.getText().toString();
            String sHora = binding.etHora.getText().toString();
            String descripcion = binding.etDescripcion.getText().toString();

            DateTimeFormatter fmtFecha = DateTimeFormatter.ofPattern("dd-MM-uuuu", Locale.getDefault());
            DateTimeFormatter fmtHora  = DateTimeFormatter.ofPattern("HH:mm",      Locale.getDefault());

            LocalDate fecha;
            LocalTime hora;

            try {
                fecha = LocalDate.parse(sFecha, fmtFecha);
                hora  = LocalTime.parse(sHora,  fmtHora);
                // usar fecha y hora aquí
            } catch (DateTimeParseException e) {
                binding.etFecha.setError("Formato de fecha inválido (dd-MM-aaaa)");
                binding.etHora.setError("Formato de hora inválido (HH:mm)");
                return;
            }

            listaEventos.add(new Evento(titulo, fecha, hora, descripcion));

            adapter.notifyItemInserted(listaEventos.size() - 1);
            binding.recyclerViewTareas.scrollToPosition(listaEventos.size() - 1);
            // Notificar al adaptador que los datos han cambiado
            adapter.notifyDataSetChanged();

            //Limpiar los editText
            binding.etTitulo.setText("");
            binding.etFecha.setText("");
            binding.etHora.setText("");
            binding.etDescripcion.setText("");
            view.clearFocus();

            Toast.makeText(MainActivity.this, "El evento se ha registrado.", Toast.LENGTH_SHORT).show();
        });

        // Inicializar RecyclerView
        initRecyclerView();

    }

    private void initRecyclerView() {
        RecyclerView recyclerView = findViewById(R.id.recyclerViewTareas);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        listaEventos = new ArrayList<>();
        adapter = new EventoAdapter(listaEventos);
        recyclerView.setAdapter(adapter);

        ExtendedFloatingActionButton botonFiltrar = binding.btnFiltrar;

        botonFiltrar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, FiltrosActivity.class);
                intent.putParcelableArrayListExtra("CLAVE_TAREAS", listaEventos);
                startActivity(intent);
            }

        });
    }

}

File: app/src/main/java/dev/ycosorio/agenda/MainActivity.kt
package dev.ycosorio.agenda

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.LinearLayoutManager
import dev.ycosorio.agenda.databinding.ActivityMainBinding
import java.time.LocalDate
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.*

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var adapter: EventoAdapter2
    private lateinit var listaEventos: ArrayList<Evento>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { view, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            view.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        initRecyclerView()
        configurarListeners()
    }

    private fun configurarListeners() {
        // LAMBDA FUNCTION para botón guardar
        binding.btnGuardar.setOnClickListener { view ->
            val titulo = binding.etTitulo.text.toString()
            val sFecha = binding.etFecha.text.toString()
            val sHora = binding.etHora.text.toString()
            val descripcion = binding.etDescripcion.text.toString()

            // Usando Kotlin Standard Functions
            val resultado = validarYCrearEvento(titulo, sFecha, sHora, descripcion)
                .also { resultado ->
                    // ALSO: Se ejecuta siempre, útil para logging
                    println("Resultado de validación: $resultado")
                }

            when (resultado) {
                is ResultadoEvento.Exito -> {
                    resultado.evento.let { nuevoEvento ->
                        // LET: Ejecuta solo si evento no es null
                        listaEventos.add(nuevoEvento)
                        adapter.notifyItemInserted(listaEventos.size - 1)
                        binding.recyclerViewTareas.scrollToPosition(listaEventos.size - 1)

                        limpiarFormulario()
                        view.clearFocus()
                        Toast.makeText(this, "Evento registrado exitosamente", Toast.LENGTH_SHORT).show()
                    }
                }
                is ResultadoEvento.Error -> {
                    Toast.makeText(this, resultado.mensaje, Toast.LENGTH_SHORT).show()
                }
            }
        }

        // ARROW FUNCTION para botón filtrar
        binding.btnFiltrar.setOnClickListener {
            Intent(this, FiltrosActivity::class.java).apply {
                // APPLY: Configura el Intent
                putParcelableArrayListExtra(FiltrosActivity.CLAVE_TAREAS, listaEventos)
            }.also { intent ->
                // ALSO: Ejecuta acción adicional
                startActivity(intent)
            }
        }
    }

    private fun initRecyclerView() {
        listaEventos = ArrayList()
        adapter = EventoAdapter2(listaEventos)

        // APPLY para configurar RecyclerView
        binding.recyclerViewTareas.apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            adapter = this@MainActivity.adapter
        }
    }

    private fun validarYCrearEvento(
        titulo: String,
        sFecha: String,
        sHora: String,
        descripcion: String
    ): ResultadoEvento {
        return run {
            // RUN: Ejecuta bloque de código y retorna el resultado
            if (titulo.trim().isEmpty()) {
                return@run ResultadoEvento.Error("El título es obligatorio")
            }

            val fmtFecha = DateTimeFormatter.ofPattern("dd-MM-yyyy", Locale.getDefault())
            val fmtHora = DateTimeFormatter.ofPattern("HH:mm", Locale.getDefault())

            try {
                val fecha = LocalDate.parse(sFecha, fmtFecha)
                val hora = LocalTime.parse(sHora, fmtHora)

                // Usando Companion Object de Evento
                val nuevoEvento = Evento.crearEventoCompleto(
                    titulo = titulo,
                    fecha = fecha,
                    hora = hora,
                    descripcion = descripcion
                )

                ResultadoEvento.Exito(nuevoEvento)
            } catch (e: DateTimeParseException) {
                with(binding) {
                    // WITH: Usar múltiples propiedades del objeto
                    etFecha.error = "Formato inválido (dd-MM-yyyy)"
                    etHora.error = "Formato inválido (HH:mm)"
                }
                ResultadoEvento.Error("Formato de fecha u hora inválido")
            }
        }
    }

    private fun limpiarFormulario() {
        // APPLY para limpiar todos los campos
        binding.apply {
            etTitulo.setText("")
            etFecha.setText("")
            etHora.setText("")
            etDescripcion.setText("")
        }
    }

    // Sealed class para manejo de resultados (característica de Kotlin)
    sealed class ResultadoEvento {
        data class Exito(val evento: Evento) : ResultadoEvento()
        data class Error(val mensaje: String) : ResultadoEvento()
    }
}